shell/utils/resourceWatcher.js

```diff
@@ -0,0 +1,212 @@
+  REQUESTED:  'requested', // requested but not confirmed by the socket yet
+  WATCHING:   'watching', // confirmed as active by the socket
+  REFRESHING: 'refreshing', // temporarily stopped while we make a full request to the API, will be rewatched afterwards
+  STOPPED:    'stopped', // temporarily stopped via message from the socket, a watch should immediately be triggered but the maintenance cycle will pick it up if that doesn't happen.
+  CANCELLED:  'cancelled', // has been flagged to send a stop request to the socket
+  REMOVED:    'removed' // stop request has been sent to the socket or it's been stopped by the socket itself and is now awaiting cleanup by the maintenanceInterval
+};
+
+export default class ResourceWatcher extends Socket {
+  watches = {};
+  maintenanceInterval;
+  connectionMetadata;
+  status = ''
+
+  constructor(url, autoReconnect = true, frameTimeout = null, protocol = null, maxTries = null) {
+    super(url, autoReconnect, frameTimeout, protocol, maxTries);
+    this.baseUrl = self.location.origin + url.replace('subscribe', '');
+
+    this.maintenanceInterval = setInterval(() => {
+      // Every 1 second we:
+      if (this.baseUrl) {
+        this.syncWatches();
+      }
+    }, 1000);
+  }
+
+  setConnectionMetadata(connectionMetadata) {
+    if (!!this.connectionMetadata) {
+      this.connectionMetadata = connectionMetadata;
+    }
+  }
+
+  async watch(watchKey, providedResourceVersion, providedResourceVersionTime, providedKeyParts = {}, providedSkipResourceVersion) {
+    const {
+      PENDING, REQUESTED, WATCHING, REFRESHING
+    } = WATCHSTATUSES;
+    const {
+      resourceType: providedResourceType,
+      id: providedId,
+      namespace: providedNamespace,
+      selector: providedSelector
+    } = providedKeyParts;
+    const resourceType = providedResourceType || this.watches?.[watchKey]?.resourceType;
+    const id = providedId || this.watches?.[watchKey]?.id;
+    const namespace = providedNamespace || this.watches?.[watchKey]?.namespace;
+    const selector = providedSelector || this.watches?.[watchKey]?.selector;
+    const resourceStatus = this.watches?.[watchKey]?.status;
+    const skipResourceVersion = this.watches?.[watchKey]?.skipResourceVersion || providedSkipResourceVersion;
+    const resourceUrl = this.baseUrl + resourceType;
+    const limitedResourceUrl = addParam(resourceUrl, 'limit', 1);
+    const opt = {
+      method:  'get',
+      headers: { accept: 'application/json' }
+    };
+    const watchObject = {
+      resourceType,
+      id,
+      namespace,
+      selector
+    };
+
+    let resourceVersionTime = providedResourceVersionTime || this.watches?.[watchKey]?.resourceVersionTime;
+    let resourceVersion = providedResourceVersion || this.watches?.[watchKey]?.resourceVersion;
+
+    if (!skipResourceVersion && (!resourceVersion || Date.now() - resourceVersionTime > 300000)) { // 300000ms is 5minutes
+      await fetch(limitedResourceUrl, opt)
```

Comment:
This skips boilerplate stuff with have in $axios, specifically the csrf header. Needs investigating why this doesn't fail

Reply:
it doesn't fail because fetch automatically inserts a cookie header which I verified by logging calls outside of the dev-tools (for whatever reason, devtools doesn't like showing the headers properly here). If need be, I can inject the CSRF cookie value into the worker from the UI thread but I can't directly get to the cookie from inside of the worker which means I'll always have to push it in. That being said, the value only changes when the session changes which would typically mean the page is at least refreshed which means I'm rebuilding the worker anyways so keeping it synced shouldn't be too big a deal.